feat(Algebra/Homology): add Euler characteristic definition for chain complexes

jessealama opened the pull request:
This PR adds a definition of the Euler characteristic for chain complexes.

The definition computes the Euler characteristic up to index n as the alternating sum ∑_{k=0}^n (-1)^k rank(C_k). It works over any ring R (using `Module.finrank`).

I'm interested in feedback on whether this is the right level of generality and whether the parameterization by index `n` is the preferred approach. This definition is extracted from work on PR #29639 (Euler's polyhedron formula).

This is a minimal PR containing only the definition - I thought it might be useful to discuss the design before adding lemmas and theorems.

joelriou commented on Mathlib/Algebra/Homology/HomologicalComplex.lean:
```diff
@@ -1013,3 +1015,16 @@ theorem mkHom_f_succ_succ (n : ℕ) :
+This is most commonly used when the base ring is a field (where rank means dimension)
+or a principal ideal domain. For a chain complex that is zero above some index,
+choosing `n` to be at least that index gives the total Euler characteristic. -/
+noncomputable def eulerChar {R : Type*} [Ring R] (C : ChainComplex (ModuleCat R) ℕ)
+    (n : ℕ) : ℤ :=
+  ∑ k ∈ Finset.range (n + 1), (-1 : ℤ)^k * Module.finrank R (C.X k)
```
This should be moved to a new file.

jessealama replied:
I was a bit uncertain where to put it. I'll put it in its own file.

jessealama commented:
I'm not quite sure what the [build error](https://github.com/leanprover-community/mathlib4/actions/runs/17716762127/job/50342742036?pr=29646) is saying. Things seem to be clean, but there's an unexplained failure at the `Check {Mathlib, Tactic, Counterexamples, Archive}.lean` step.

joelriou commented on Mathlib/Algebra/Homology/EulerCharacteristic.lean:
```diff
@@ -0,0 +1,34 @@
+This is most commonly used when the base ring is a field (where rank means dimension)
+or a principal ideal domain. For a chain complex that is zero above some index,
+choosing `n` to be at least that index gives the total Euler characteristic. -/
+noncomputable def eulerChar (C : ChainComplex (ModuleCat R) ℕ) (n : ℕ) : ℤ :=
+  ∑ k ∈ Finset.range (n + 1), (-1 : ℤ)^k * Module.finrank R (C.X k)
```
```suggestion
noncomputable def eulerChar (C : ChainComplex (ModuleCat R) ℕ) (n : ℕ) : ℤ :=
  ∑ k ∈ Finset.range n, (-1 : ℤ)^k * Module.finrank R (C.X k)
```
In order to do inductions, it may be easier to start from the `0` instead of `1`. Could you add lemmas `eulerChar_zero`, `eulerChar_succ`?

jessealama replied:
I've added an unbounded Euler characteristic and added a theorem that relates the bounded and unbounded versions. Thanks!

jessealama committed:
feat(Algebra/Homology): add Euler characteristic for chain complexes

Thanks @joelriou for suggesting a number of improvements!

This introduces the Euler characteristic for chain complexes:
- `boundedEulerChar`: The finite alternating sum ∑_{k=0}^{n-1} (-1)^k rank(C_k)
- `eulerChar`: The infinite alternating sum ∑_{k=0}^∞ (-1)^k rank(C_k) using tsum

Key lemmas:
- `eulerChar_zero`: The bounded characteristic at 0 is 0 (empty sum)
- `eulerChar_succ`: Recursive formula for the bounded characteristic
- `eulerChar_eq_boundedEulerChar`: The two definitions agree when the complex eventually vanishes

Co-authored-by: Joël Riou <joel.riou@universite-paris-saclay.fr>

jessealama commented:
Definitions generalized and moved to #29713
